File: Python/Relativity_Modules_SSS_mods/Relativity_Modules_SSS_mod6.0_just__solve_ivp/shared/potential_extrema.py
class PotentialDerivative:
    """
    A class for computing dV/dR in a scenario-agnostic way.
    It takes scenario-specific functions f and f1 (passed from main.py).
    """

    def __init__(self, f, f1):
        """
        Initialize the class with scenario-specific f and f1.

        :param f: A callable representing f(R)
        :param f1: A callable representing f'(R)
        """
        self.f = f
        self.f1 = f1

    def dVdR(self, R):
        """
        Computes dV/dR = (1/3) * (2*f(R) - R*f1(R))
        as described in the paper.

        :param R: The Ricci scalar
        :return: The value of dV/dR at R
        """
        return (1/3) * (2 * self.f(R) - self.f1(R) * R)

import numpy as np
from scipy.optimize import brentq


def find_roots(f, x_start, x_end, num_points=500):
    x_grid = np.linspace(x_start, x_end, num_points)
    roots = []

    for i in range(len(x_grid) - 1):
        x_left = x_grid[i]
        x_right = x_grid[i+1]
        f_left = f(x_left)
        f_right = f(x_right)

        # Check if either endpoint is effectively zero
        tol = 1e-12
        if abs(f_left) < tol:
            roots.append(x_left)
            continue
        if abs(f_right) < tol:
            roots.append(x_right)
            continue

        # Check for sign change
        if f_left * f_right < 0:
            try:
                root = brentq(f, x_left, x_right)
                roots.append(root)
            except ValueError:
                # This would be unusual if there's a genuine sign change
                pass

    return roots

File: Julia/[0] relativity in Julia/shared/potential_extrema.jl
module PotentialExtrema

using Roots

export PotentialDerivative, dVdR, find_roots

struct PotentialDerivative
    f::Function
    f1::Function
end

function dVdR(pd::PotentialDerivative, R)
    """
    dV/dR = (1/3) * (2*f(R) - R*f1(R))
    """
    return (1/3) * (2 * pd.f(R) - pd.f1(R) * R)
end

function find_roots(f, x_start, x_end, num_points=500)
    x_grid = range(x_start, x_end, length=num_points)
    roots = Float64[]
    for i in 1:length(x_grid)-1
        x_left = x_grid[i]
        x_right = x_grid[i+1]
        f_left = f(x_left)
        f_right = f(x_right)
        tol = 1e-12
        if abs(f_left) < tol
            push!(roots, x_left)
            continue
        end
        if abs(f_right) < tol
            push!(roots, x_right)
            continue
        end
        if f_left * f_right < 0
            try
                root = find_zero(f, (x_left, x_right), Bisection())
                push!(roots, root)
            catch e
                # If no root is found, continue silently.
            end
        end
    end
    return roots
end

end  # module PotentialExtrema
